File: Cpp/src/problems/puzzle8.cpp
#include "puzzle8.h"

#include <algorithm>
#include <iostream>

std::shared_ptr<Node<BoardState> > Puzzle8::_moveUp(std::shared_ptr<Node<BoardState> > node)
{
    auto emptyTile = std::find(node->state.begin(), node->state.end(), 0);

    int emptyIndex = std::distance(node->state.begin(), emptyTile);
    int row        = emptyIndex / ROW_SIZE;

    // if the tile is in the first row
    if (row == 0) return nullptr;

    std::shared_ptr<Node<BoardState> > newNode = std::make_shared<Node<BoardState> >(*node);
    newNode->parent                            = node;
    newNode->cost++;

    std::swap(newNode->state[emptyIndex], newNode->state[emptyIndex - ROW_SIZE]);

    return newNode;
}

std::shared_ptr<Node<BoardState> > Puzzle8::_moveDown(std::shared_ptr<Node<BoardState> > node)
{
    auto emptyTile = std::find(node->state.begin(), node->state.end(), 0);

    int emptyIndex = std::distance(node->state.begin(), emptyTile);
    int row        = emptyIndex / ROW_SIZE;

    // if the tile is in the last row
    if (row == ROW_SIZE - 1) return nullptr;

    std::shared_ptr<Node<BoardState> > newNode = std::make_shared<Node<BoardState> >(*node);
    newNode->parent                            = node;
    newNode->cost++;

    std::swap(newNode->state[emptyIndex], newNode->state[emptyIndex + ROW_SIZE]);

    return newNode;
}

std::shared_ptr<Node<BoardState> > Puzzle8::_moveLeft(std::shared_ptr<Node<BoardState> > node)
{
    auto emptyTile = std::find(node->state.begin(), node->state.end(), 0);

    int emptyIndex = std::distance(node->state.begin(), emptyTile);
    int col        = emptyIndex % COL_SIZE;

    // if the tile is in the first col
    if (col == 0) return nullptr;

    std::shared_ptr<Node<BoardState> > newNode = std::make_shared<Node<BoardState> >(*node);
    newNode->parent                            = node;
    newNode->cost++;

    std::swap(newNode->state[emptyIndex], newNode->state[emptyIndex - 1]);

    return newNode;
}

std::shared_ptr<Node<BoardState> > Puzzle8::_moveRight(std::shared_ptr<Node<BoardState> > node)
{
    auto emptyTile = std::find(node->state.begin(), node->state.end(), 0);

    int emptyIndex = std::distance(node->state.begin(), emptyTile);
    int col        = emptyIndex % COL_SIZE;

    // if the tile is in the last col
    if (col == COL_SIZE - 1) return nullptr;

    std::shared_ptr<Node<BoardState> > newNode = std::make_shared<Node<BoardState> >(*node);
    newNode->parent                            = node;
    newNode->cost++;

    std::swap(newNode->state[emptyIndex], newNode->state[emptyIndex + 1]);

    return newNode;
}

void Puzzle8::_printBoard(BoardState &board)
{
    for (int i = 0; i < COL_SIZE; ++i) {
        for (int j = 0; j < ROW_SIZE; ++j) {
            std::cout << board[i * ROW_SIZE + j] << " ";
        }
        std::cout << "\n";
    }
}

Puzzle8::Puzzle8(BoardState initialState, BoardState goalState)
    : Problem<BoardState>(initialState, goalState)
{
}

bool Puzzle8::isGoal(std::shared_ptr<Node<BoardState> > node)
{
    if (node->state == _goalState) {
        _goalNode = std::make_shared<Node<BoardState> >(node);
        return true;
    }
    return false;
}

std::vector<std::shared_ptr<Node<BoardState> > > Puzzle8::expand(
    std::shared_ptr<Node<BoardState> > node)
{
    std::vector<std::shared_ptr<Node<BoardState> > > expandedNodes;

    if (auto upNode = _moveUp(node)) {
        expandedNodes.push_back(upNode);
    }

    if (auto downNode = _moveDown(node)) {
        expandedNodes.push_back(downNode);
    }

    if (auto leftNode = _moveLeft(node)) {
        expandedNodes.push_back(leftNode);
    }

    if (auto rightNode = _moveRight(node)) {
        expandedNodes.push_back(rightNode);
    }

    return expandedNodes;
}

void Puzzle8::printSolution()
{
    auto node = _goalNode;

    if (!node) {
        std::cout << "No solution found.\n";
        return;
    }

    std::vector<std::shared_ptr<Node<BoardState> > > path;

    while (node) {
        path.push_back(node);
        node = node->parent;
    }

    std::reverse(path.begin(), path.end());

    for (const auto &step : path) {
        _printBoard(step->state);
        std::cout << "-----\n";
    }
}

File: src/problems/Puzzle8.py
from searchAlgorithms.helpers.HeuristicProblem import HeuristicProblem
from searchAlgorithms.helpers.Node import Node


class Puzzle8(HeuristicProblem):
    ROW_SIZE = 3
    COL_SIZE = 3

    def __init__(self, initialState: list, goalState: list):
        super().__init__(initialState, goalState)

        self._goalNode: Node | None = None

    def _moveUp(self, node: Node):
        emptyIndex = node.state.index(0)
        row = emptyIndex // self.ROW_SIZE

        # if the tile is in the first row
        if row == 0:
            return None

        newNode = Node(node.state[:], node)
        newNode.cost = node.cost + 1

        newNode.state[emptyIndex], newNode.state[emptyIndex - self.ROW_SIZE] = (
            newNode.state[emptyIndex - self.ROW_SIZE],
            newNode.state[emptyIndex],
        )

        return newNode

    def _moveDown(self, node: Node):
        emptyIndex = node.state.index(0)
        row = emptyIndex // self.ROW_SIZE

        # if the tile is in the last row
        if row == self.ROW_SIZE - 1:
            return None

        newNode = Node(node.state[:], node)
        newNode.cost = node.cost + 1

        newNode.state[emptyIndex], newNode.state[emptyIndex + self.ROW_SIZE] = (
            newNode.state[emptyIndex + self.ROW_SIZE],
            newNode.state[emptyIndex],
        )

        return newNode

    def _moveLeft(self, node: Node):
        emptyIndex = node.state.index(0)
        col = emptyIndex % self.COL_SIZE

        # if the tile is in the first col
        if col == 0:
            return None

        newNode = Node(node.state[:], node)
        newNode.cost = node.cost + 1

        newNode.state[emptyIndex], newNode.state[emptyIndex - 1] = (
            newNode.state[emptyIndex - 1],
            newNode.state[emptyIndex],
        )

        return newNode

    def _moveRight(self, node: Node):
        emptyIndex = node.state.index(0)
        col = emptyIndex % self.COL_SIZE

        # if the tile is in the first col
        if col == self.COL_SIZE - 1:
            return None

        newNode = Node(node.state[:], node)
        newNode.cost = node.cost + 1

        newNode.state[emptyIndex], newNode.state[emptyIndex + 1] = (
            newNode.state[emptyIndex + 1],
            newNode.state[emptyIndex],
        )

        return newNode

    @property
    def initialNodes(self):
        return [self._initialNode]

    @property
    def goalNode(self) -> Node | None:
        return self._goalNode

    def isGoal(self, node: Node):
        if node.state == self._goalState:
            self._goalNode = node
            return True
        return False

    def expand(self, node: Node) -> list[Node]:
        expandedNodes = []

        upNode = self._moveUp(node)
        if upNode:
            expandedNodes.append(upNode)

        downNode = self._moveDown(node)
        if downNode:
            expandedNodes.append(downNode)

        leftNode = self._moveLeft(node)
        if leftNode:
            expandedNodes.append(leftNode)

        rightNode = self._moveRight(node)
        if rightNode:
            expandedNodes.append(rightNode)

        return expandedNodes

    def hashableState(self, node: Node):
        return tuple(node.state)

    def heuristic(self, node: Node):
        # currState: List = node.state
        # value = 0
        #
        # for i, state in enumerate(currState):
        #     if state != self._goalState[i]:
        #         value += 1
        #
        # return value
        currState = node.state
        distance = 0

        for i, tile in enumerate(currState):
            if tile == 0:
                continue

            goalIndex = self._goalState.index(tile)
            x1, y1 = divmod(i, 3)
            x2, y2 = divmod(goalIndex, 3)
            distance += abs(x1 - x2) + abs(y1 - y2)

        return distance
